packages/forms-web-app/src/dynamic-forms/validator/date-validator.js

```diff
@@ -0,0 +1,153 @@
+		this.noDayMonthErrorMessage =
+			errorMessages?.noDayMonthErrorMessage ?? defaultErrorMessages.noDayMonthErrorMessage;
+		this.noDayYearErrorMessage =
+			errorMessages?.noDayYearErrorMessage ?? defaultErrorMessages.noDayYearErrorMessage;
+		this.noMonthYearErrorMessage =
+			errorMessages?.noMonthYearErrorMessage ?? defaultErrorMessages.noMonthYearErrorMessage;
+		this.invalidDateErrorMessage =
+			errorMessages?.invalidDateErrorMessage ?? defaultErrorMessages.invalidDateErrorMessage;
+		this.invalidYearErrorMessage =
+			errorMessages?.invalidYearErrorMessage ?? defaultErrorMessages.invalidYearErrorMessage;
+		this.futureDateErrorMessage =
+			errorMessages?.futureDateErrorMessage ?? defaultErrorMessages.futureDateErrorMessage;
+	}
+
+	/**
+	 * validates the response body, checking the values sent for the date are valid
+	 * @param {Question} questionObj
+	 */
+	validate(questionObj) {
+		const fieldName = questionObj.fieldName;
+		const dayInput = `${fieldName}_day`;
+		const monthInput = `${fieldName}_month`;
+		const yearInput = `${fieldName}_year`;
+
+		return [
+			// check all or some date inputs are not empty
+			body(dayInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (!req.body[monthInput] && !req.body[yearInput]) {
+						return this.emptyErrorMessage;
+					}
+
+					if (!req.body[monthInput] && req.body[yearInput]) {
+						return this.noDayMonthErrorMessage;
+					}
+
+					if (req.body[monthInput] && !req.body[yearInput]) {
+						return this.noDayYearErrorMessage;
+					}
+
+					return this.noDayErrorMessage;
+				}),
+			body(monthInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (req.body[dayInput] && !req.body[yearInput]) {
+						return this.noMonthYearErrorMessage;
+					}
+					if (req.body[dayInput]) {
+						return this.noMonthErrorMessage;
+					}
+				}),
+			body(yearInput)
+				.notEmpty()
+				.withMessage((_, { req }) => {
+					if (req.body[dayInput] && req.body[monthInput]) return this.noYearErrorMessage;
+				}),
+
+			// check date values entered are within valid ranges
+			body(dayInput)
+				.isInt({ min: 1, max: 31 })
+				.withMessage(this.invalidDateErrorMessage)
+				.bail()
+				.toInt()
+				.custom((value, { req }) => {
+					const monthNum = parseInt(req.body[monthInput], 10);
+					if (value === 31 && monthNum && [4, 6, 9, 11].includes(monthNum)) return false;
+
+					if (value > 28 && monthNum && monthNum === 2) {
+						const yearNum = parseInt(req.body[yearInput], 10);
+						return yearNum && isLeapYear(new Date(yearNum, 1, 1)) && value === 29;
+					}
+
+					return true;
+				})
+				.withMessage(this.invalidDateErrorMessage),
+			body(monthInput).isInt({ min: 1, max: 12 }).withMessage(this.invalidDateErrorMessage),
+			body(yearInput).isInt({ min: 1000, max: 9999 }).withMessage(this.invalidYearErrorMessage),
+
+			//check date is not in the future
+			body(dayInput).custom((value, { req }) => {
```

Comment:
We might need to be able to configure if the validation is a check on the future or the past, I think there is another question where it checks the date is allowed to be in the past but not the future

Reply:
as discussed, we do not need to implement this functionality for this ticket - the existing functionality can be adapted in a future ticket when requirements demand.